File: sdks/python/mailcue/resources/mailboxes.py
"""``mailboxes`` resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from mailcue.resources._base import AsyncResource, SyncResource
from mailcue.types import (
    EmailListResponse,
    Mailbox,
    MailboxListResponse,
    MailboxStats,
)


def _create_payload(
    username: str,
    password: str,
    *,
    domain: Optional[str],
    display_name: str,
) -> Dict[str, Any]:
    payload: Dict[str, Any] = {
        "username": username,
        "password": password,
        "display_name": display_name,
    }
    if domain is not None:
        payload["domain"] = domain
    return payload


def _list_emails_params(
    folder: str,
    page: int,
    page_size: int,
    search: Optional[str],
    sort: str,
) -> Dict[str, Any]:
    params: Dict[str, Any] = {
        "folder": folder,
        "page": page,
        "page_size": page_size,
        "sort": sort,
    }
    if search is not None:
        params["search"] = search
    return params


class Mailboxes(SyncResource):
    """Synchronous ``mailboxes`` resource."""

    def list(self) -> MailboxListResponse:
        """List all mailboxes.

        Example:
            >>> client.mailboxes.list()
        """
        response = self._transport.request("GET", "/mailboxes")
        return MailboxListResponse.model_validate(response.json())

    def create(
        self,
        username: str,
        password: str,
        *,
        domain: Optional[str] = None,
        display_name: str = "",
    ) -> Mailbox:
        """Create a new mailbox.

        Example:
            >>> client.mailboxes.create("alice", "s3cret", domain="example.com")
        """
        payload = _create_payload(username, password, domain=domain, display_name=display_name)
        response = self._transport.request("POST", "/mailboxes", json=payload)
        return Mailbox.model_validate(response.json())

    def delete(self, address: str) -> None:
        """Delete a mailbox by address.

        Example:
            >>> client.mailboxes.delete("alice@example.com")
        """
        self._transport.request("DELETE", f"/mailboxes/{address}")

    def stats(self, mailbox_id: str) -> MailboxStats:
        """Return IMAP STATUS counts for a mailbox.

        Example:
            >>> client.mailboxes.stats(mailbox_id="abc")
        """
        response = self._transport.request("GET", f"/mailboxes/{mailbox_id}/stats")
        return MailboxStats.model_validate(response.json())

    def purge(self, address: str) -> Dict[str, Any]:
        """Permanently delete every email in a mailbox.

        Example:
            >>> client.mailboxes.purge("alice@example.com")
        """
        response = self._transport.request("POST", f"/mailboxes/{address}/purge")
        result: Dict[str, Any] = response.json() if response.content else {}
        return result

    def list_emails(
        self,
        address: str,
        *,
        folder: str = "INBOX",
        page: int = 1,
        page_size: int = 50,
        search: Optional[str] = None,
        sort: str = "date_desc",
    ) -> EmailListResponse:
        """List emails in a mailbox (mailbox-scoped variant of ``emails.list``).

        Example:
            >>> client.mailboxes.list_emails("alice@example.com")
        """
        params = _list_emails_params(folder, page, page_size, search, sort)
        response = self._transport.request("GET", f"/mailboxes/{address}/emails", params=params)
        return EmailListResponse.model_validate(response.json())


class AsyncMailboxes(AsyncResource):
    """Asynchronous ``mailboxes`` resource."""

    async def list(self) -> MailboxListResponse:
        """Async variant of :meth:`Mailboxes.list`."""
        response = await self._transport.request("GET", "/mailboxes")
        return MailboxListResponse.model_validate(response.json())

    async def create(
        self,
        username: str,
        password: str,
        *,
        domain: Optional[str] = None,
        display_name: str = "",
    ) -> Mailbox:
        """Async variant of :meth:`Mailboxes.create`."""
        payload = _create_payload(username, password, domain=domain, display_name=display_name)
        response = await self._transport.request("POST", "/mailboxes", json=payload)
        return Mailbox.model_validate(response.json())

    async def delete(self, address: str) -> None:
        """Async variant of :meth:`Mailboxes.delete`."""
        await self._transport.request("DELETE", f"/mailboxes/{address}")

    async def stats(self, mailbox_id: str) -> MailboxStats:
        """Async variant of :meth:`Mailboxes.stats`."""
        response = await self._transport.request("GET", f"/mailboxes/{mailbox_id}/stats")
        return MailboxStats.model_validate(response.json())

    async def purge(self, address: str) -> Dict[str, Any]:
        """Async variant of :meth:`Mailboxes.purge`."""
        response = await self._transport.request("POST", f"/mailboxes/{address}/purge")
        result: Dict[str, Any] = response.json() if response.content else {}
        return result

    async def list_emails(
        self,
        address: str,
        *,
        folder: str = "INBOX",
        page: int = 1,
        page_size: int = 50,
        search: Optional[str] = None,
        sort: str = "date_desc",
    ) -> EmailListResponse:
        """Async variant of :meth:`Mailboxes.list_emails`."""
        params = _list_emails_params(folder, page, page_size, search, sort)
        response = await self._transport.request(
            "GET", f"/mailboxes/{address}/emails", params=params
        )
        return EmailListResponse.model_validate(response.json())

File: sdks/node/src/resources/mailboxes.ts
import type { Transport } from '../transport.js';
import { camelize } from '../transport.js';
import type {
  EmailListResponse,
  ListEmailsParams,
  Mailbox,
  MailboxCreateParams,
  MailboxListResponse,
  MailboxStats,
} from '../types.js';

type Opts = { signal?: AbortSignal };

export class MailboxesResource {
  constructor(private readonly transport: Transport) {}

  async list(options: Opts = {}): Promise<MailboxListResponse> {
    const reqOpts: Parameters<Transport['request']>[0] = {
      method: 'GET',
      path: '/api/v1/mailboxes',
    };
    if (options.signal) reqOpts.signal = options.signal;
    const raw = await this.transport.request<unknown>(reqOpts);
    return camelize(raw) as MailboxListResponse;
  }

  async create(params: MailboxCreateParams, options: Opts = {}): Promise<Mailbox> {
    const body: Record<string, unknown> = {
      username: params.username,
      password: params.password,
      display_name: params.displayName ?? '',
    };
    if (params.domain !== undefined) body['domain'] = params.domain;
    const reqOpts: Parameters<Transport['request']>[0] = {
      method: 'POST',
      path: '/api/v1/mailboxes',
      body,
    };
    if (options.signal) reqOpts.signal = options.signal;
    const raw = await this.transport.request<unknown>(reqOpts);
    return camelize(raw) as Mailbox;
  }

  async delete(address: string, options: Opts = {}): Promise<void> {
    const reqOpts: Parameters<Transport['request']>[0] = {
      method: 'DELETE',
      path: `/api/v1/mailboxes/${encodeURIComponent(address)}`,
      responseType: 'void',
    };
    if (options.signal) reqOpts.signal = options.signal;
    await this.transport.request<void>(reqOpts);
  }

  async stats(mailboxId: string, options: Opts = {}): Promise<MailboxStats> {
    const reqOpts: Parameters<Transport['request']>[0] = {
      method: 'GET',
      path: `/api/v1/mailboxes/${encodeURIComponent(mailboxId)}/stats`,
    };
    if (options.signal) reqOpts.signal = options.signal;
    const raw = await this.transport.request<unknown>(reqOpts);
    return camelize(raw) as MailboxStats;
  }

  async purge(address: string, options: Opts = {}): Promise<void> {
    const reqOpts: Parameters<Transport['request']>[0] = {
      method: 'POST',
      path: `/api/v1/mailboxes/${encodeURIComponent(address)}/purge`,
      responseType: 'void',
    };
    if (options.signal) reqOpts.signal = options.signal;
    await this.transport.request<void>(reqOpts);
  }

  async listEmails(
    address: string,
    params: Omit<ListEmailsParams, 'mailbox'> = {},
    options: Opts = {},
  ): Promise<EmailListResponse> {
    const query: Record<string, unknown> = {};
    if (params.folder !== undefined) query['folder'] = params.folder;
    if (params.page !== undefined) query['page'] = params.page;
    if (params.pageSize !== undefined) query['page_size'] = params.pageSize;
    if (params.search !== undefined) query['search'] = params.search;
    if (params.sort !== undefined) query['sort'] = params.sort;
    const reqOpts: Parameters<Transport['request']>[0] = {
      method: 'GET',
      path: `/api/v1/mailboxes/${encodeURIComponent(address)}/emails`,
      query,
    };
    if (options.signal) reqOpts.signal = options.signal;
    const raw = await this.transport.request<unknown>(reqOpts);
    return camelize(raw) as EmailListResponse;
  }
}
